### PR TITLE
Exclude substitution definitions from gettext build (take 3)

### PR DESCRIPTION
**Pull request by alvinhochun**
### Feature or Bugfix

- Bugfix

### Purpose
- Fix `rst_prolog` and `rst_epilog` substitutions appearing in every POT files pointlessly.

### Detail
- Exclude substitution definitions from the gettext message extraction, to avoid substitution definitions from appearing in every POT files even if the substitutions are not used in those pages. Fixes a regression from ebc3c68b78d5cf51f9aa49ca38dd371b7c204cd8 which was introduced in 3.5.0.

### Relates
- #9428

Supersedes #9629

cc @therahedwig

### CODE REVIEW

File: sphinx/builders/gettext.py [hunk 1]
```diff
@@ -144,6 +144,17 @@ def compile_catalogs(self, catalogs: Set[CatalogInfo], message: str) -> None:
...
+        def is_node_part_of_substitution_definition(node: nodes.Node) -> bool:
+            """Check the node and its parents to see if any of them is a substitution
+            definition.
+            """
+            # return False
+            if isinstance(node, nodes.substitution_definition):
+                return True
+            if node.parent:
+                return is_node_part_of_substitution_definition(node.parent)
+            return False
```
Comment (AA-Turner) [inline sphinx/builders/gettext.py]:
```suggestion
            while node.parent:
                if isinstance(node, nodes.substitution_definition):
                    return True
                node = node.parent
            return False
```

File: sphinx/builders/gettext.py [hunk 2]
```diff
@@ -144,6 +144,17 @@ def compile_catalogs(self, catalogs: Set[CatalogInfo], message: str) -> None:
         return
 
     def write_doc(self, docname: str, doctree: nodes.document) -> None:
+        def is_node_part_of_substitution_definition(node: nodes.Node) -> bool:
```
Comment (AA-Turner) [inline sphinx/builders/gettext.py]:
```suggestion
        def is_node_in_substitution_definition(node: nodes.Node) -> bool:
```

File: sphinx/builders/gettext.py [hunk 3]
```diff
@@ -152,6 +163,8 @@ def write_doc(self, docname: str, doctree: nodes.document) -> None:
                 catalog.add(msg, node)
 
         for node, msg in extract_messages(doctree):
+            if is_node_part_of_substitution_definition(node):
```
Comment (AA-Turner) [inline sphinx/builders/gettext.py]:
```suggestion
            if is_node_in_substitution_definition(node):
```

File: sphinx/environment/collectors/asset.py [hunk 4]
```diff
@@ -79,8 +79,11 @@ def process_doc(self, app: Sphinx, doctree: nodes.document) -> None:
 
                 # Update `node['uri']` to a relative path from srcdir
                 # from a relative path from current document.
+                original_uri = node['uri']
                 node['uri'], _ = app.env.relfn2path(imguri, docname)
                 candidates['*'] = node['uri']
+                if node['uri'] != original_uri:
+                    node['original_uri'] = original_uri
```
Comment (AA-Turner) [inline sphinx/environment/collectors/asset.py:78]: @alvinhochun -- is this needed for this PR? If not, could it be in a different PR please?

A

File: sphinx/builders/gettext.py [hunk 5]
```diff
@@ -148,6 +148,17 @@ def compile_catalogs(self, catalogs: set[CatalogInfo], message: str) -> None:
         return
 
     def write_doc(self, docname: str, doctree: nodes.document) -> None:
+        def is_node_in_substitution_definition(node: nodes.Node) -> bool:
+            """Check the node and its parents to see if any of them is a substitution
+            definition.
+            """
+            # return False
```
Comment (AA-Turner) [inline sphinx/builders/gettext.py]:
```suggestion
```

File: tests/test_build_gettext.py [hunk 6]
```diff
@@ -202,3 +202,73 @@ def test_build_single_pot(app):
...
+
+    def msgid_getter(msgid):
+        m = _msgid_getter(msgid)
+        if m:
+            return m.groups()[0]
+        return None
+
+    assert (app.outdir / 'prolog_epilog_substitution.pot').isfile()
+    pot = (app.outdir / 'prolog_epilog_substitution.pot').read_text()
+    msgids = [_f for _f in map(msgid_getter, pot.splitlines()) if _f]
```
Comment (AA-Turner) [inline tests/test_build_gettext.py]:
```suggestion
    msgids = list(filter(None, map(msgid_getter, pot.splitlines())))
```

File: tests/test_build_gettext.py [hunk 7]
```diff
@@ -202,3 +202,73 @@ def test_build_single_pot(app):
...
+
+    _msgid_getter = re.compile(r'msgid "(.*)"').search
+
+    def msgid_getter(msgid):
+        m = _msgid_getter(msgid)
+        if m:
+            return m.groups()[0]
+        return None
+
+    assert (app.outdir / 'prolog_epilog_substitution.pot').isfile()
```
Comment (AA-Turner) [inline tests/test_build_gettext.py]:
```suggestion
    assert os.path.isfile(app.outdir / 'prolog_epilog_substitution.pot')
```

File: sphinx/environment/collectors/asset.py (hunk 4 above)
Comment (alvinhochun) [reply sphinx/environment/collectors/asset.py:78]: Hi, I don't exactly remember the context for this commit, if that change make sense to you would you mind splitting that commit into its own PR for me? Otherwise you can try to revert or remove that commit from this PR.

(FWIW it *might* have been related to image URIs inside substitutions, but I'm just guessing from the test cases. See if the tests pass without this commit.)

Comment (alvinhochun) [reply sphinx/environment/collectors/asset.py:78]: The context might be [this comment](https://github.com/sphinx-doc/sphinx/pull/9454#issuecomment-884172757).

**Comment by AA-Turner** [comment]
Thanks @alvinhochun!

A
